### ISSUE TITLE
Incorrect "position-only" error in dunder method definitions when parameter name starts with `__`

### ISSUE THREAD

**Issue by aholmes** [microsoft/pyright#7359]
In defining a class used as a metaclass, I define several dunder methods that are assigned to a new type created during runtime. A real example of this can be found [here](https://github.com/uclahs-cds/BL_Python/blob/aholmes-add-postgresql-support/src/programming/BL_Python/programming/patterns/singleton.py#L79-L125).

When I define these "nested" dunder methods using parameters that begin with `__`, pyright reports

> error: Position-only parameter not allowed after parameter that is not position-only (reportGeneralTypeIssues)

The error is resolved by removing the preceding `__`.

The issue can be reproduced with this:

```python
from typing import Any
class Foo:
    def __new__(cls: type, cls_name: str, bases: tuple[type[Any]], members: dict[str, Any]):
        def __init__(self: Any, __x: str) -> None: ...
        def __getattribute__(self: Any, __name: str) -> Any: ...
```

pyright version 1.1.351

$ pyright foo.py
foo.py
  foo.py:8:33 - error: Position-only parameter not allowed after parameter that is not position-only (reportGeneralTypeIssues)
  foo.py:10:41 - error: Position-only parameter not allowed after parameter that is not position-only (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations

**Comment by erictraut** [comment]
Yes, this is correct behavior. In this context, `__init__` and `__getattribute__` are not methods because they're not defined in a class body. They're simple functions, and `self` and `cls` have no special meaning in this context. In non-method functions, position-only parameters cannot follow non-position-only parameters regardless of whether you're using the new-style `/` parameter separator introduced in [PEP 570](https://peps.python.org/pep-0570/) or the older-style double-underscore naming convention. This was recently clarified in [the typing spec](https://typing.readthedocs.io/en/latest/spec/historical.html#positional-only-parameters). 

You have a couple of options here:
1. Rename `self` and `cls` to `__self` and `__cls`.
2. Abandon the use of the old (defunct) mechanism and remove the double underscore parameter names. Then use the newer PEP 570 notation and add a `/` separator after the last parameter to indicate that all preceding parameters should be treated as position-only.

```python
from typing import Any


class Foo:
    def __new__(
        cls: type, cls_name: str, bases: tuple[type[Any]], members: dict[str, Any]
    ):
        def __init__(self: Any, x: str, /) -> None: ...
        def __getattribute__(self: Any, name: str, /) -> Any: ...
```

**Comment by aholmes** [comment]
>  the older-style double-underscore naming convention. This was recently clarified in [the typing spec](https://typing.readthedocs.io/en/latest/spec/historical.html#positional-only-parameters).

Ah ha, thank you. I had tried to find whether this was expected but didn't see this. Appreciate the reference.
